File: app/src/main/java/app/android/ttp/mikazuki/yoshinani/services/GroupService.java
package app.android.ttp.mikazuki.yoshinani.services;

import android.content.Context;
import android.support.annotation.NonNull;

import org.greenrobot.eventbus.EventBus;

import app.android.ttp.mikazuki.yoshinani.event.FetchDataEvent;
import app.android.ttp.mikazuki.yoshinani.event.FetchListDataEvent;
import app.android.ttp.mikazuki.yoshinani.event.RefreshEvent;
import app.android.ttp.mikazuki.yoshinani.model.GroupModel;
import app.android.ttp.mikazuki.yoshinani.model.GroupUserModel;
import app.android.ttp.mikazuki.yoshinani.repository.retrofit.ApiUtil;
import app.android.ttp.mikazuki.yoshinani.repository.retrofit.service.RetrofitGroupService;
import rx.Subscription;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;

/**
 * @author haijimakazuki
 */
public class GroupService implements Subscription {

    final private String TAG = this.getClass().getSimpleName();
    private final EventBus eventbus = EventBus.getDefault();
    RetrofitGroupService mAPI;
    private Context mContext = null;

    public GroupService(Context context) {
        this.mContext = context;
        mAPI = ApiUtil
                .buildRESTAdapter(mContext)
                .create(RetrofitGroupService.class);
    }

    public void getAll(final int userId) {
        mAPI.getAll(userId)
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(response -> eventbus.post(new FetchListDataEvent<>(GroupModel.from(response.body())))
                        , throwable -> eventbus.post(new FetchListDataEvent<>(null)));
    }

    public void get(final int groupId) {
        mAPI.get(groupId)
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(response -> eventbus.post(new FetchDataEvent<>(GroupModel.from(response.body())))
                        , throwable -> eventbus.post(new FetchDataEvent<>(null)));
    }

    public void create(@NonNull final GroupModel group) {
        mAPI.create(new RetrofitGroupService.RequestDataOnCreate(group))
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(response -> eventbus.post(new RefreshEvent())
                        , throwable -> eventbus.post(new FetchDataEvent<>(null)));
    }

    public void update(@NonNull final GroupModel group) {
        mAPI.update(group.getId(), new RetrofitGroupService.RequestDataOnCreate(group))
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(response -> eventbus.post(new FetchDataEvent<>(GroupModel.from(response.body())))
                        , throwable -> eventbus.post(new FetchDataEvent<>(null)));
    }

    public void invite(int groupId, int userId) {
        mAPI.invite(groupId, new RetrofitGroupService.RequestDataOnInvite(userId))
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(response -> eventbus.post(new FetchDataEvent<>(GroupModel.from(response.body())))
                        , throwable -> eventbus.post(new FetchDataEvent<>(null)));
    }

    public void accept(int groupId) {
        mAPI.accept(groupId)
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(response -> eventbus.post(new FetchDataEvent<>(GroupUserModel.from(response.body())))
                        , throwable -> eventbus.post(new FetchDataEvent<>(null)));
    }

    @Override
    public void unsubscribe() {

    }

    @Override
    public boolean isUnsubscribed() {
        return false;
    }
}

File: app/src/main/java/app/android/ttp/mikazuki/yoshinani/services/GroupService.kt
package app.android.ttp.mikazuki.yoshinani.services

import android.content.Context

import org.greenrobot.eventbus.EventBus

import app.android.ttp.mikazuki.yoshinani.event.FetchDataEvent
import app.android.ttp.mikazuki.yoshinani.event.FetchListDataEvent
import app.android.ttp.mikazuki.yoshinani.event.RefreshEvent
import app.android.ttp.mikazuki.yoshinani.model.GroupModel
import app.android.ttp.mikazuki.yoshinani.model.GroupUserModel
import app.android.ttp.mikazuki.yoshinani.repository.retrofit.ApiUtil
import app.android.ttp.mikazuki.yoshinani.repository.retrofit.service.RetrofitGroupService
import rx.Subscription
import rx.android.schedulers.AndroidSchedulers
import rx.schedulers.Schedulers

/**
 * @author haijimakazuki
 */
class GroupService(context: Context) : Subscription {

    private val TAG = this.javaClass.getSimpleName()
    private val eventbus = EventBus.getDefault()
    internal var mAPI: RetrofitGroupService
    private val mContext: Context

    init {
        this.mContext = context
        mAPI = ApiUtil
                .buildRESTAdapter(mContext)
                .create(RetrofitGroupService::class.java!!)
    }

    fun getAll(userId: Int) {
        mAPI.getAll(userId)
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ response -> eventbus.post(FetchListDataEvent(GroupModel.from(response.body())!!)) }, { throwable -> eventbus.post(FetchListDataEvent<Any>(null!!)) })
    }

    operator fun get(groupId: Int) {
        mAPI.get(groupId)
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ response -> eventbus.post(FetchDataEvent(GroupModel.from(response.body()))) }, { throwable -> eventbus.post(FetchDataEvent<Any>(null)) })
    }

    fun create(group: GroupModel) {
        mAPI.create(RetrofitGroupService.RequestDataOnCreate(group))
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ response -> eventbus.post(RefreshEvent()) }, { throwable -> eventbus.post(FetchDataEvent<Any>(null)) })
    }

    fun update(group: GroupModel) {
        mAPI.update(group.id, RetrofitGroupService.RequestDataOnCreate(group))
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ response -> eventbus.post(FetchDataEvent(GroupModel.from(response.body()))) }, { throwable -> eventbus.post(FetchDataEvent<Any>(null)) })
    }

    fun invite(groupId: Int, userId: Int) {
        mAPI.invite(groupId, RetrofitGroupService.RequestDataOnInvite(userId))
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ response -> eventbus.post(FetchDataEvent(GroupModel.from(response.body()))) }, { throwable -> eventbus.post(FetchDataEvent<Any>(null)) })
    }

    fun accept(groupId: Int) {
        mAPI.accept(groupId)
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe({ response -> eventbus.post(FetchDataEvent(GroupUserModel.from(response.body()))) }, { throwable -> eventbus.post(FetchDataEvent<Any>(null)) })
    }

    override fun unsubscribe() {

    }

    override fun isUnsubscribed(): Boolean {
        return false
    }
}
